docs/design/2018-07-01-refactor-aggregate-framework.md

```diff
@@ -0,0 +1,72 @@
+- That the partial result can be any type also means the aggregate functions can use a map with the specific input type as keys. For example, use `map[types.MyDecimal]` to de-duplicate the input decimal values. In this way, the overhead on the encoding and decoding operations on the old framework can be reduced.
+- `UpdatePartialResult` is called with a batch of input records. The overhead caused by the per record function call on the framework can be saved. Since all the execution operators are using `Chunk` to store the input rows, in which the data belonging to the same column is stored consecutively in the memory, the aggregate functions should be executed one by one, take full utilization of the CPU caches, reduce the cache misses, and improve the execution performance.
+- For every state and every kind of input type, a specific aggregate function should be implemented to handle it. This means the CPU branch predictions on the aggregate state and input value types can reduce during the execution of `UpdatePartialResult`, utilize the CPU pipelines, and improve the execution speed.
+- `AppendFinalResult2Chunk` directly finalizes the partial result to the chunk, without converting it to `Datum` and then converting the `Datum` back into `Chunk`. This saves a lot of object allocation, reduces the overhead of golang’s gc worker, and avoids the unnecessary value convertings between `Datum` and `Chunk`.
+
+Disadvantages:
+- An aggregate function need to be implemented for every possible state and input type. This could introduces a lot of development work. And it takes more coding work to add a new aggregate function.
+
+## Compatibility
+
+For now, the new framework is only supported in the stream aggregate operator. If the `Build()` function returns a `nil` pointer, it rolls back to the old framework.
+
+So this new framework can be tested during the development, and all the results should be the same with the old framework.
+
+## Implementation
+
+- Implement the new framework, including the `AggFunc` interface and the `Build()` function. [@zz-jason](https://github.com/zz-jason)
+- Implement the `AVG` function under the new framework. [@zz-jason](https://github.com/zz-jason)
+- Implement other aggregate functions under the new framework. [@XuHuaiyu](https://github.com/XuHuaiyu)
+- Add more integration tests using the random query generator. [@XuHuaiyu](https://github.com/XuHuaiyu)
+- Add unit tests for the new framework. [@XuHuaiyu](https://github.com/XuHuaiyu)
+- Infer input parameter type for new aggregate functions when extracting a `Project` operator under the `Aggregate` operator. [@XuHuaiyu](https://github.com/XuHuaiyu)
+- Use the new framework in the hash aggregate operator. [@XuHuaiyu](https://github.com/XuHuaiyu)
+
+## Open issues (if applicable)
+
+- https://github.com/pingcap/tidb/issues/7042: add unit tests for package "aggfuncs".
```

Comment:
According to the template, `Open issue` is for the issues that we don't have a solution.